truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/TruffleInstrument.java

```diff
@@ -0,0 +1,299 @@
+         * @throws IllegalStateException if the method is called later than from
+         *             {@link #onCreate(com.oracle.truffle.api.instrumentation.TruffleInstrument.Env) }
+         *             method
+         */
+        public void registerService(Object service) {
+            if (services == null) {
+                throw new IllegalStateException();
+            }
+            services.add(service);
+        }
+
+        Object[] onCreate(TruffleInstrument instrumentation) {
+            List<Object> arr = new ArrayList<>();
+            services = arr;
+            try {
+                instrumentation.onCreate(this);
+            } finally {
+                services = null;
+            }
+            return arr.toArray();
+        }
+
+        /**
+         * Evaluates source of (potentially different) language using the current context.The names
+         * of arguments are parameters for the resulting {#link CallTarget} that allow the
+         * <code>source</code> to reference the actual parameters passed to
+         * {@link CallTarget#call(java.lang.Object...)}.
+         *
+         * @param source the source to evaluate
+         * @param argumentNames the names of {@link CallTarget#call(java.lang.Object...)} arguments
+         *            that can be referenced from the source
+         * @return the call target representing the parsed result
+         * @throws IOException if the parsing or evaluation fails for some reason
+         */
+        @SuppressWarnings("static-method")
+        public CallTarget parse(Source source, String... argumentNames) throws IOException {
+            return InstrumentationHandler.ACCESSOR.parse(null, source, null, argumentNames);
+        }
+
+    }
+
+    /**
+     * Annotation to register an {@link TruffleInstrument instrument} implementations for automatic
+     * discovery.
+     */
+    @Retention(RetentionPolicy.SOURCE)
+    @Target(ElementType.TYPE)
+    public @interface Registration {
+
+        /**
+         * A custom machine identifier for this instrumentation. If not defined then the fully
+         * qualified class name is used.
+         */
+        String id() default "";
+
+        /**
+         * The name of the instrumentation in an arbitrary format for humans.
+         */
+        String name() default "";
+
+        /**
+         * The version for instrumentation in an arbitrary format.
+         */
+        String version() default "";
+
+        /**
+         * Whether or not the annotated instrumentation should be enabled by default. If autostart
+         * is <code>true</code> then {@link TruffleInstrument#onCreate(Env)} is invoked for this
+         * instrument just before the first guest language code is executed. If autostart is
+         * <code>false</code> then the instrument needs to be enabled using the engine.
+         */
+        boolean autostart() default false;
+
```

Comment:
Thank you! Yes, I also want to say again, that I am against opt-out in general. I do prefer very much an opt-in model, per `PolyglotEngine` instance.

Not only that I dislike random jars on the classpath potentially breaking my carefully built interpreter, I also find it counter intuitive to have this code everywhere were I create a PolyglotEngine to get the desired semantics: `engine.getInstruments().values().forEach(i -> i.setEnabled(false));`
Especially since I do not know what all the instruments are that might be autostarted.